S3C-1985: Reduce memory usage of listing algorithms

DavidPineauScality opened the pull request:
The issue discovered in the field was that an key with a heavy amount of data
could lead to an exhaustion of the available memory for Node.JS javascript
code.
In the case of the buckets handling, that would be MPU objects with an
important number of parts, listed in the "location" field.

A customer case revealed and triggered this corner case, loop-crashing the
processes responsible for the listing of the database, which were using the
Arsenal.algos.list.* extensions.

That issue is hereby fixed by this commit through the trimming of the known
heavy unused data: the "location" field of the bucket's values.
Note that as the code previously didn't need to parse the value before
forwarding it back to the caller, we are now parsing it, removing the unwanted
fields, and re-stringifying it for each selected entry in the listing.

A notable impact of this change is that the CPU usage might go up substantially
while listing the contents of any bucket.
Additionally, and for safety purposes, if the data cannot be parsed and
alterated, it will be returned as-is, at the risk of maintaining the memory
consuming behavior in the case of that corner case; while an error is logged on
stderr.

Fixes S3C-1985

Commit:
S3C-1985: Reduce memory usage of listing algorithms

The issue discovered in the field was that an key with a heavy amount of data
could lead to an exhaustion of the available memory for Node.JS javascript
code.
In the case of the buckets handling, that would be MPU objects with an
important number of parts, listed in the "location" field.

A customer case revealed and triggered this corner case, loop-crashing the
processes responsible for the listing of the database, which were using the
Arsenal.algos.list.* extensions.

That issue is hereby fixed by this commit through the trimming of the known
heavy unused data: the "location" field of the bucket's values.
Note that as the code previously didn't need to parse the value before
forwarding it back to the caller, we are now parsing it, removing the unwanted
fields, and re-stringifying it for each selected entry in the listing.

A notable impact of this change is that the CPU usage might go up substantially
while listing the contents of any bucket.
Additionally, and for safety purposes, if the data cannot be parsed and
alterated, it will be returned as-is, at the risk of maintaining the memory
consuming behavior in the case of that corner case; while a warning is logged
on stdout.

Fixes S3C-1985

bert-e commented:
# Hello davidpineauscality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/arsenal/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/S3C-1985/listing-filter-value` into
`stabilization/7.4.3`
* [w/7.4/bugfix/S3C-1985/listing-filter-value](https://github.com/scality/arsenal/compare/development/7.4...w/7.4/bugfix/S3C-1985/listing-filter-value) will be merged into `development/7.4` (pull request #667)
* [w/7.5/bugfix/S3C-1985/listing-filter-value](https://github.com/scality/arsenal/compare/development/7.5...w/7.5/bugfix/S3C-1985/listing-filter-value) will be merged into `development/7.5` (pull request #668)
* [w/8.0/bugfix/S3C-1985/listing-filter-value](https://github.com/scality/arsenal/compare/development/8.0...w/8.0/bugfix/S3C-1985/listing-filter-value) will be merged into `development/8.0` (pull request #669)
* [w/8.1/bugfix/S3C-1985/listing-filter-value](https://github.com/scality/arsenal/compare/development/8.1...w/8.1/bugfix/S3C-1985/listing-filter-value) will be merged into `development/8.1` (pull request #670)



The following branches will **NOT** be impacted:

* `development/6.4`




*Follow* integration pull requests if you would like to be notified of
build statuses by email.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers





Peer approvals *must* include at least 1 approval from the following list:

* @rachedbenmustapha

* @rahulreddy

* @bennettbuchanan

DavidPineauScality commented on lib/algos/list/Extension.js:
```diff
@@ -23,6 +23,32 @@ class Extension {
+     *
+     * @return {Object} The value without the filtered-out fields.
+     */
+    filterContent(value) {
```
I was hesitant about adding a code comment explaining the issue here (similarly to the commit message).
I could also add a message hinting to read the commit message associated to the addition of the function.

Please tell me what you think would be best (or do the associated edit yourself, either way :) )

Commit:
S3C-1985: Fix tests after memory consumption fix

In the associated issues's memory consumption fix, a warning is logged in case
of unparseable entry from the listing.

This broke the unit tests for some specific listing algorithms, as they had
never taken any logger as constructor parameter, as opposed to their base
class.

Taking a look at the usage of these classes in the known client code
(CloudServer & Metadata), the logger was actually always provided, which means
that we could make use of it, and accept this forgotten parameter.

Fixes tests for S3C-1985, Fixes silent prototype usage mismatch.

ghost commented:
@bert-e force_reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/S3C-1985/listing-filter-value` into
`stabilization/7.4.3`
* [w/7.4/bugfix/S3C-1985/listing-filter-value](https://github.com/scality/arsenal/compare/development/7.4...w/7.4/bugfix/S3C-1985/listing-filter-value) will be merged into `development/7.4` (pull request #671)
* [w/7.5/bugfix/S3C-1985/listing-filter-value](https://github.com/scality/arsenal/compare/development/7.5...w/7.5/bugfix/S3C-1985/listing-filter-value) will be merged into `development/7.5` (pull request #672)
* [w/8.0/bugfix/S3C-1985/listing-filter-value](https://github.com/scality/arsenal/compare/development/8.0...w/8.0/bugfix/S3C-1985/listing-filter-value) will be merged into `development/8.0` (pull request #673)
* [w/8.1/bugfix/S3C-1985/listing-filter-value](https://github.com/scality/arsenal/compare/development/8.1...w/8.1/bugfix/S3C-1985/listing-filter-value) will be merged into `development/8.1` (pull request #674)



The following branches will **NOT** be impacted:

* `development/6.4`




*Follow* integration pull requests if you would like to be notified of
build statuses by email.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers





Peer approvals *must* include at least 1 approval from the following list:

* @rachedbenmustapha

* @rahulreddy

* @bennettbuchanan

DavidPineauScality commented on lib/algos/list/delimiter.js:
```diff
@@ -61,8 +61,8 @@ class Delimiter extends Extension {
      */
-    constructor(parameters) {
-        super(parameters);
+    constructor(parameters, logger) {
```
As described in commit number 2, I had to add this for tests compatibility.
As the loggers are actually provided by cloudserver & metadata, there should be no issue here, and we should have the logger to log in the base Extension's `filterContent()` method.

bert-e commented:
# History mismatch



Merge commit #53520f4d7ea3af6974366b0f698f80a7d29904d4 on the integration branch
`w/7.4/bugfix/S3C-1985/listing-filter-value` is merging a branch which is neither the current
branch `bugfix/S3C-1985/listing-filter-value` nor the development branch
`development/7.4`.

It is likely due to a rebase of the branch `bugfix/S3C-1985/listing-filter-value` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

ghost commented:
@bert-e force_reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/S3C-1985/listing-filter-value` into
`stabilization/7.4.3`
* [w/7.4/bugfix/S3C-1985/listing-filter-value](https://github.com/scality/arsenal/compare/development/7.4...w/7.4/bugfix/S3C-1985/listing-filter-value) will be merged into `development/7.4` (pull request #675)
* [w/7.5/bugfix/S3C-1985/listing-filter-value](https://github.com/scality/arsenal/compare/development/7.5...w/7.5/bugfix/S3C-1985/listing-filter-value) will be merged into `development/7.5` (pull request #676)
* [w/8.0/bugfix/S3C-1985/listing-filter-value](https://github.com/scality/arsenal/compare/development/8.0...w/8.0/bugfix/S3C-1985/listing-filter-value) will be merged into `development/8.0` (pull request #677)
* [w/8.1/bugfix/S3C-1985/listing-filter-value](https://github.com/scality/arsenal/compare/development/8.1...w/8.1/bugfix/S3C-1985/listing-filter-value) will be merged into `development/8.1` (pull request #678)



The following branches will **NOT** be impacted:

* `development/6.4`




*Follow* integration pull requests if you would like to be notified of
build statuses by email.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers





Peer approvals *must* include at least 1 approval from the following list:

* @rachedbenmustapha

* @rahulreddy

* @bennettbuchanan

DavidPineauScality commented on lib/algos/list/Extension.js:
```diff
@@ -23,6 +23,32 @@ class Extension {
+            delete ret.location;
+            ret = JSON.stringify(ret);
+        } catch (e) {
+            this.logger.warn('Could not parse Object Metadata while listing',
```
If deemed necessary, we could protect against an undefined logger here with a conditional construct.
Wasn't sure it'd be worth it, so just mentionning it here.

jonathan-gramain commented on lib/algos/list/Extension.js:
```diff
@@ -23,6 +23,32 @@ class Extension {
+    filterContent(value) {
+        let ret = undefined;
+        try {
+            ret = JSON.parse(value);
```
As an improvement attempt to limit the CPU cost I would prefer to have a heuristic condition on the blob size before executing this, so the location field will be trimmed only if the blob is really big. Then we pay the serialize/deserialize cost only when we're certain to gain a significant number of bytes in the result (only if our assumption that the "location" field takes up the bulk of the blob bytes, which in theory should be the case).

This should amortize the cost only to memory-greedy listings and not have an impact on the vast majority of listings, considering the blob size check to be much cheaper than the serialization cycle.

DavidPineauScality replied:
Good point. LGTM, that was one of my worries

jonathan-gramain commented on lib/algos/list/Extension.js:
```diff
@@ -23,6 +23,32 @@ class Extension {
+     *
+     * @return {Object} The value without the filtered-out fields.
+     */
+    filterContent(value) {
```
minor: renaming to `trimContents` would be more precise I think

DavidPineauScality replied:
ok for me

jonathan-gramain reviewed (commented):
I will attempt the changes I suggested and test on my lab.

jonathan-gramain committed:
S3C-1985: heuristic to limit metadata trim to large blob

Add a heuristic to only do the trim for large metadata blob of more
than 10KB. This should limit the number of JSON (de)serializations to
only those blobs where we can hope a significant reduction in size.

Also renamed "filterContent" to "trimMetadata"

bert-e commented:
# History mismatch



Merge commit #3c448ff9944a4f18ecb95c0ca00ea6d2c2acf0de on the integration branch
`w/7.4/bugfix/S3C-1985/listing-filter-value` is merging a branch which is neither the current
branch `bugfix/S3C-1985/listing-filter-value` nor the development branch
`development/7.4`.

It is likely due to a rebase of the branch `bugfix/S3C-1985/listing-filter-value` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

jonathan-gramain reviewed (approved):
Please someone validates my changes as well

vrancurel reviewed (approved):
Ideally we should strip down any sub-document which is larger than a treshold, e.g. in some cases user-metadata or tags can also be > 10K, and therefore we should also strip them down.

jonathan-gramain commented:
So we would have to extend the `delete` list to those potentially large attributes too when we're in the 10K+ case? Why not, just we would have to do the thorough analysis that it does not break any code that might need them (same kind than has been done for the location field already). Not sure it's worth it for now IMHO but open to discussion.

vrancurel commented:
> Not sure it's worth it for now IMHO but open to discussion.

This is why I approved the PR :)

vrancurel commented:
The correct way to implement this would have been acc/to me to have the listing returning only the minimal attributes needed for supporting the S3 listing (e.g. lastModified and maybe others). And do an optional params.projection API typical of document stores where consumers would typically asks for specific fields to be returned. I know it might be useful one day to be able to retrieve sub-documents while listing. I am not aware of an existing use case right now. Also we would have needed to change cloudserver+bucketclient+bucketd+repd which would have been complicated.

This is why also I approved this PR.

jonathan-gramain commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/S3C-1985/listing-filter-value` into
`stabilization/7.4.3`
* [w/7.4/bugfix/S3C-1985/listing-filter-value](https://github.com/scality/arsenal/compare/development/7.4...w/7.4/bugfix/S3C-1985/listing-filter-value) will be merged into `development/7.4` (pull request #680)
* [w/7.5/bugfix/S3C-1985/listing-filter-value](https://github.com/scality/arsenal/compare/development/7.5...w/7.5/bugfix/S3C-1985/listing-filter-value) will be merged into `development/7.5` (pull request #681)
* [w/8.0/bugfix/S3C-1985/listing-filter-value](https://github.com/scality/arsenal/compare/development/8.0...w/8.0/bugfix/S3C-1985/listing-filter-value) will be merged into `development/8.0` (pull request #682)
* [w/8.1/bugfix/S3C-1985/listing-filter-value](https://github.com/scality/arsenal/compare/development/8.1...w/8.1/bugfix/S3C-1985/listing-filter-value) will be merged into `development/8.1` (pull request #683)



The following branches will **NOT** be impacted:

* `development/6.4`




*Follow* integration pull requests if you would like to be notified of
build statuses by email.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers





Peer approvals *must* include at least 1 approval from the following list:

* @rachedbenmustapha

* @rahulreddy

* @bennettbuchanan

DavidPineauScality commented on lib/algos/list/Extension.js:
```diff
@@ -30,22 +47,28 @@ class Extension {
+                // Prefer returning an unfiltered data rather than
+                // stopping the service in case of parsing failure.
+                // The risk of this approach is a potential
+                // reproduction of MD-692, where too much memory is
```
Might need rewording into S3C-1985, though I think the link MD-692 should still be valid.

rahulreddy reviewed (approved):
LGTM. So we are not going ahead with the optional params idea?

ghost commented:
@rahulreddy Indeed, going from the principle that the field which was an actual issue was never used in CloudServer, as opposed to the suspect one we were looking at (partsLocation) when devising this idea.

ghost commented:
Given that both Tests on MetaData and CloudServer are green with this patch (branch bugfix/S3C-1985-runTestsWithArsenalPatch , if you're looking for it in the builds listings); I'm considering this patch as OK, given the approval of both @jonathan-gramain and @rahulreddy.

ghost commented:
@bert-e approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `stabilization/7.4.3`

* :heavy_check_mark: `development/7.4`

* :heavy_check_mark: `development/7.5`

* :heavy_check_mark: `development/8.0`

* :heavy_check_mark: `development/8.1`



The following branches will **NOT** be impacted:

* `development/6.4`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `stabilization/7.4.3`

* :heavy_check_mark: `development/7.4`

* :heavy_check_mark: `development/7.5`

* :heavy_check_mark: `development/8.0`

* :heavy_check_mark: `development/8.1`



The following branches have **NOT** changed:

* `development/6.4`



Please check the status of the associated issue S3C-1985.

Goodbye davidpineauscality.

ghost commented:
the PR was stuck on the tests of one integration branch which were not properly triggered.
I forced them, to unblock the situation.
